[2.3] Static Code Analysis for Components

kalessil opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Static Code Analysis with Php Inspections (EA Extended), no functional changes:
- array_keys/array_values usage as foreach array
- foreach value by reference - added unsets to keep scope clear
- strstr usage as strpos fixed
- array_push miss-use resolved

kalessil commented:
@nicolas-grekas : as proposed <a href="https://github.com/symfony/symfony/pull/13995#discussion_r26871175">here</a> foreach related issues cleaned up.

src/Symfony/Component/Security/Acl/Dbal/AclProvider.php:650
src/Symfony/Component/Security/Acl/Dbal/AclProvider.php:658

I can not follow what happens there, so both loops remain untouched.

fabpot commented:
ping @nicolas-grekas

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Debug/TraceableEventDispatcher.php:
```diff
@@ -472,11 +472,13 @@ private function unwrapListener($listener, $eventId)
-                foreach (array_keys($this->wrappedListeners) as $eventId) {
-                    if (isset($this->wrappedListeners[$eventId][$listener])) {
-                        return $this->wrappedListeners[$eventId][$listener];
+                foreach ($this->wrappedListeners as $eventId => &$eventListeners) {
```
Since the function does not return by ref, this ref should be removed

kalessil replied:
Removed. Since $eventListeners is not a reference anymore, complementary unset is also removed.

nicolas-grekas commented:
:+1: with comment above to be fixed before merging

stof commented on src/Symfony/Component/Console/Descriptor/ApplicationDescription.php:
```diff
@@ -144,9 +144,11 @@ private function sortCommands(array $commands)
+            ksort($commandsSet);
         }
+        // unset reference to keep scope clear
+        unset($commandsSet);
```
this is not necessary. It will be removed anyway when reaching the end of the method on next line

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToArrayTransformer.php:
```diff
@@ -99,6 +99,8 @@ public function transform($dateTime)
                 $entry = (string) (int) $entry;
             }
+            // unset reference to keep scope clear
+            unset($entry);
```
is there any value about doing this ?

nicolas-grekas replied:
If the $entry var is reused later in the same scope, this ensures that the ref won't be reused by mistake. This is not necessary, but it's still cautious, thus good to me. Removing as much refs as possible would be even better, but I think the remaining cases are justified.

kalessil replied:
@stof : this has sense only when foreach value variable is a reference. 
As function grows, variable (reference) got reused and "ghost" array change bug might happen.

kalessil commented:
Shall not be merged until additional clarifications.

kalessil commented:
Travis still crashing due to memcached installation failures.

kalessil committed:
Php Inspections (EA Extended): static code analysis

- array_keys/array_values usage as foreach array
- foreach value by reference
- added unsets
- strstr usage as strpos fixed
- array_push miss-use resolved

kalessil committed:
Php Inspections (EA Extended): added remark regarding php fatal found by new plugin version

kalessil commented:
ping @nicolas-grekas

nicolas-grekas commented:
:+1:

jderusse commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php:
```diff
@@ -39,7 +39,7 @@ public function process(ContainerBuilder $container)
+        foreach ($container->getDefinitions() as $id => $definition) {
             // yes, we are specifically fetching the definition from the
             // container to ensure we are not operating on stale data
             $definition = $container->getDefinition($id);
```
Is this line still necessary?

stof replied:
@jderusse yes, because the loop could change other definitions. the comment is still accurate

fabpot commented:
Thank you @kalessil.
